src/Backend.idr

```diff
@@ -12,86 +12,49 @@ import Text.PrettyPrint.WL
-  constructor MkDecl
-
-  ||| The name of the type.
-  name   : Name
-
-  ||| The names of the type's parameters.
-  params : Vect n Name -- TODO is number of parameters enough?
-
-||| A variable environment. Left=free, Right=bound.
-Env : Nat -> Type
-Env k = Vect k (Either String Decl)
-
-||| Standard implmementation of `freshEnv` for languages that require type parameters to
-||| start with lower case letters (and allow numbers in names).
-freshEnvLC : (n : Nat) -> Env n
-freshEnvLC n = unindex {n} (\f => Left ("x" ++ show (finToInteger f)))
-
-||| Standard implmementation of `freshEnv` for languages that require type parameters to
-||| start with upper case letters (and allow numbers in names).
-freshEnvUC : (n : Nat) -> Env n
-freshEnvUC n = unindex {n} (\f => Left ("X" ++ show (finToInteger f)))
-
-||| Get the free names from the environment.
-getFreeVars : (e : Env n) -> Vect (fst (Vect.filter Either.isLeft e)) String
-getFreeVars e with (filter isLeft e)
-  | (p ** v) = map (either id (const "")) v
-
-||| Get a list of the de Brujin indices that are actually used in a `TDef`.
-getUsedIndices : TDef n -> List (Fin n)
-getUsedIndices T0         = []
-getUsedIndices T1         = []
-getUsedIndices (TSum xs)  = assert_total $ concatMap getUsedIndices xs
-getUsedIndices (TProd xs) = assert_total $ concatMap getUsedIndices xs
-getUsedIndices (TVar i)   = [i]
-getUsedIndices (TMu _ xs) = assert_total $ concatMap ((concatMap weedOutZero) . getUsedIndices . snd) xs
-  where weedOutZero : Fin (S n) -> List (Fin n)
-        weedOutZero FZ     = []
-        weedOutZero (FS i) = [i]
-getUsedIndices (TName _ t) = getUsedIndices t
-
-||| Filter out the entries in an `Env` that is referred to by a `TDef`.
-getUsedVars : Env n -> (td: TDef n) -> Env (length (getUsedIndices td))
-getUsedVars e td = map (flip index e) (fromList $ getUsedIndices td)
+||| Interface for interpreting type definitions as ASTs.
+||| @def  the type representing type definitions.
+||| @type the type representing type signatures.
+||| @n    the number of variables the interpretor supports in a type definition. (Should always be either `n` or `0`.)
+interface ASTGen def type (n : Nat) | def where
+  ||| Given a `TDef`, generate its corresponding type signature.
+  msgType        : TNamed n -> type
 
-||| Interface for codegens. `lang` is a type representing (the syntactic structure of)
-||| a type declaration in the target language.
-interface Backend lang where
-  ||| Given a `TDef` and a matching environment, generate a list of type definitions
-  ||| in the target language.
-  generateTyDefs : Env n -> TDef n -> List lang
+  ||| Generate definitions for a `TNamed` and all its helper definitions.
+  generateTyDefs : TNamed n -> List def
 
-  ||| Given a type definition in the target language, generate its code.
-  generateCode : lang -> Doc
+||| Interface for code generators that can generate code for type definitions and
+||| type signatures independently of each other.
```

Comment:
It might be nice to list some examples here, e.g. from the PR description 
"Haskell and ReasonML".